subsystems/networking/bluetooth.go

```diff
@@ -0,0 +1,630 @@
+	return nil
+}
+
+// Stop stops advertising a bluetooth service which (when enabled) accepts WiFi and Viam cloud config credentials.
+func (bsl *bluetoothServiceLinux) stop() error {
+	bsl.mu.Lock()
+	defer bsl.mu.Unlock()
+
+	if bsl.adv == nil {
+		return errors.New("advertisement is nil")
+	}
+	if !bsl.advActive {
+		return errors.New("invalid request, advertising already inactive")
+	}
+	if err := bsl.adv.Stop(); err != nil {
+		return fmt.Errorf("failed to stop advertising: %w", err)
+	}
+	bsl.advActive = false
+	bsl.logger.Debug("Stopped advertising a BLE connection")
+	return nil
+}
+
+// WaitForCredentials returns credentials, the minimum required information to provision a robot and/or its WiFi.
+func (bsl *bluetoothServiceLinux) waitForCredentials(
+	ctx context.Context, requiresCloudCredentials, requiresWiFiCredentials bool, inputChan chan<- userInput,
+) error {
+	if !requiresWiFiCredentials && !requiresCloudCredentials {
+		return errors.New("should be waiting for cloud credentials or WiFi credentials, or both")
+	}
+	var ssid, psk, robotPartKeyID, robotPartKey string
+	var ssidErr, pskErr, robotPartKeyIDErr, robotPartKeyErr error
+	for {
+		var shouldBreakOuterLoop bool
+		if ctx.Err() != nil {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			shouldBreakOuterLoop = true
+		default:
+			if requiresWiFiCredentials {
+				if ssid == "" {
+					var e *emptyBluetoothCharacteristicError
+					ssid, ssidErr = bsl.readSsid()
+					if ssidErr != nil && !errors.As(ssidErr, &e) {
+						shouldBreakOuterLoop = true
+						break
+					}
+				}
+				if psk == "" {
+					var e *emptyBluetoothCharacteristicError
+					psk, pskErr = bsl.readPsk()
+					if pskErr != nil && !errors.As(pskErr, &e) {
+						shouldBreakOuterLoop = true
+						break
+					}
+				}
+			}
+			if requiresCloudCredentials {
+				if robotPartKeyID == "" {
+					var e *emptyBluetoothCharacteristicError
+					robotPartKeyID, robotPartKeyIDErr = bsl.readRobotPartKeyID()
+					if robotPartKeyIDErr != nil && !errors.As(robotPartKeyIDErr, &e) {
+						shouldBreakOuterLoop = true
+						break
+					}
+				}
+				if robotPartKey == "" {
+					var e *emptyBluetoothCharacteristicError
+					robotPartKey, robotPartKeyErr = bsl.readRobotPartKey()
+					if robotPartKeyErr != nil && !errors.As(robotPartKeyErr, &e) {
+						shouldBreakOuterLoop = true
+						break
+					}
+				}
+			}
+			if requiresWiFiCredentials && requiresCloudCredentials && //nolint:gocritic
+				ssid != "" && psk != "" && robotPartKeyID != "" && robotPartKey != "" {
+				shouldBreakOuterLoop = true
+				break
+			} else if requiresWiFiCredentials && ssid != "" && psk != "" {
+				shouldBreakOuterLoop = true
+				break
+			} else if requiresCloudCredentials && robotPartKeyID != "" && robotPartKey != "" {
+				shouldBreakOuterLoop = true
+				break
+			}
+
+			// Not ready to return (do not have the minimum required set of credentials), so sleep and try again.
+			time.Sleep(time.Second)
```

Comment:
This needs to integrate with healthchecks. Use the healthySleep functions there. May need to extend if the health of bluetooth is separate from the general subsystems loops.